### PR TITLE
Initial minimal UX story for GPU PCI Passthrough in TKG/CAPV

### LINKED ISSUE #2828: Expose ability to passthrough PCI devices in vSphere 7+ from CAPV v1.3.0 (TBD)
**Describe the feature request**

The next release of Cluster API Provider vSphere will include support for mapping PCI devices into [VMs using passthrough support (aka vSphere DirectPath)](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1264). We want to light this up feature in TKG, both pre and post ClusterClass.

We additionally require allowing users to change the machine deployment rollout strategy because they might not have sufficient hardware to do an upgrade otherwise.

Required UX:

Add the following variables to existing vSphere cluster plans:

### `VSPHERE_WORKER_PCI_DEVICES` and `VSPHERE_CONTROL_PLANE_PCI_DEVICES`

* Type: String
* Default: ""
* Variable to be modified:
  - `VSPHERE_WORKER_PCI_DEVICES`:
    - Modify all vSphereMachineTemplates matching `"${ CLUSTER_NAME }-worker"`
    - Modifies `spec.template.spec.pciDevices` as below
  - `VSPHERE_CONTROL_PLANE_PCI_DEVICES`:
    - Modify all vSphereMachineTemplates matching `"${ CLUSTER_NAME }-control-plane"`
    - Modifies `spec.template.spec.pciDevices` as below

If "", do not set. If value other than "", then do the following:

```psuedocode
# This is just pseudocode. Needs to be in YTT (Starlark).

devices = []
kvs = string.split(value, ",")
if len(kvs) == 0, return nil
for each kv in kvs:
  if !VSPHERE_IGNORE_PCI_DEVICES_ALLOW_LIST && kv not in allowedPCIDevices:
     return validation error
  vendor, device = string.split(kv, ":")
  if device == nil, return validation error
  devices.append(
   {
      vendorID: vendor,
      deviceID: deviceID
   }
  )
```

E.g:

Assuming the `allowedDevices` list contains `00001:10000` and `00002:20000`, then:

`VSPHERE_WORKER_PCI_DEVICES="00001:10000,00002:20000"` should result in

```yaml
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: VSphereMachineTemplate
metadata: cluster-worker
...
spec:
...
  template
    spec:
    ...
    pciDevices:
    - vendorID: 00001
      deviceID: 10000   
    - vendorID: 00002
      deviceID: 20000 
```

The following should return an "invalid format" error:
* `VSPHERE_WORKER_PCI_DEVICES="00001:10000;00002:20000"`
* `VSPHERE_WORKER_PCI_DEVICES="00001:10000,00002;20000"`
* `VSPHERE_WORKER_PCI_DEVICES="0000110000"`

The following should return a "One or more PCI devices not validated. Set VSPHERE_IGNORE_PCI_DEVICES_ALLOW_LIST to true to override":

* `VSPHERE_WORKER_PCI_DEVICES="00001:10000;00003:30000"`
* `VSPHERE_WORKER_PCI_DEVICES="00003:30000"`

### `WORKER_ROLLOUT_STRATEGY`

* Type: String
* Enum: `OnDelete` or `RollingUpdate`. Fail validation if other value provided.
* Default: `RollingUpdate`
* Modifies: Any `MachineDeployment.spec.strategy.type` to the specified value.

This setting is generic across all providers, so does not have the vSphere prefix

### `VSPHERE_CONTROL_PLANE_CUSTOM_VMX_KEYS` and `VSPHERE_WORKER_CUSTOM_VMX_KEYS`

* Type: String
* Default: ""

If set, deconstructs comma-separated key value pairs, e.g.

`VSPHERE_CUSTOM_VMX_KEYS="pciPassthru.allowP2P=true,pciPassthru.RelaxACSforP2P=true,pciPassthru.use64bitMMIO=true,pciPassthru.64bitMMIOSizeGB=512"` into a map in spec.template.spec.customVMXKeys, e.g.

```yaml
      customVMXKeys:
        pciPassthru.64bitMMIOSizeGB: "512"
```

### Stretch goal: `VSPHERE_IGNORE_PCI_DEVICES_ALLOW_LIST`

* Type: Boolean
* Default: false
* Behaviour: If set, disables the enumeration for vendor and device class in `VSPHERE_WORKER_PCI_DEVICES` and `VSPHERE_CONTROL_PLANE_PCI_DEVICES`

## Acceptance Criteria

* [vSphere templates and YTT updated to handle the logic above](https://github.com/vmware-tanzu/tanzu-framework/tree/main/pkg/v1/providers/infrastructure-vsphere)
* [ClusterGen and unit tests added where appropriate](https://github.com/vmware-tanzu/tanzu-framework/tree/main/pkg/v1/providers/tests)
* [TKG Client library for vSphere](https://github.com/vmware-tanzu/tanzu-framework/blob/main/pkg/v1/tkg/tkgconfigproviders/vsphere.go) updated to also include the variables
* Validation errors given to end user when the input is invalid

## Reviewers

For Cluster API vSphere correctness, @yastij and @randomvariable  will be reviewers
For Tanzu Framework correctness, @anujc25 

**Affected product area (please put an X in all that apply)**

- [X] APIs
- [X] CLI
- [X] Installation
- [X] Plugin
- [X] User Experience

**Additional context**

### PR DESCRIPTION
**Pull request by NEWBEE1**
### What this PR does / why we need it
Tanzu CLI changes for GPU PCI passthrough support
### Which issue(s) this PR fixes

Fixes #2828 

### Describe testing done for PR

- Unit Testing
- Manual testing to verify all the parameters are filled properly.  

### Release note

```release-note
Tanzu CLI changes for GPU PCI passthrough support
```

### PR Checklist

- [ ] Squash the commits into one or a small number of logical commits
      
- [ ] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [ ] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

**Comment by jayunit100** [comment]
THANK YOU ! looks great esp the unit test coverage ! 

squash commits + i think theres some kinda failure on the hack/verify step for formatting of the new config yaml ?

**Review by jayunit100 — APPROVED** [review]
mostly looks good, but i think need to fix the lint errors

File: pkg/v1/providers/config_default.yaml [hunk 1]
```diff
@@ -132,6 +132,15 @@ VSPHERE_DATASTORE:
 #! name of an existing VM folder in which to place Tanzu Kubernetes Grid VMs
 VSPHERE_FOLDER:
 VSPHERE_STORAGE_POLICY_ID: ""
+
+#!
```
Comment (randomvariable) [inline pkg/v1/providers/config_default.yaml:136]: Complete comment

File: pkg/v1/providers/infrastructure-azure/v1.2.1/ytt/overlay.yaml [hunk 2]
```diff
@@ -36,6 +38,14 @@
...
 #@ end
 
+#@ def verify_and_configure_strategy(strategytype):
+#@   if strategytype not in strateg_type:
+#@     assert.fail("Invalid Strategy type  \""+strategytype+"\", it must be OnDelete or RollingUpdate")
+#@   end
+#@   return strategytype
+#@ end
+
+
```
Comment (randomvariable) [inline pkg/v1/providers/infrastructure-azure/v1.2.1/ytt/overlay.yaml:48]: Dedupe please, please put this into common library in pkg/v1/providers/lib/helpers.star as

```starlark
valid_strategy_types = ["OnDelete", "RollingUpdate"]


def verify_and_configure_machine_deployment_rollout_strategy(strategy_type):
if strategy_type not in valid_strategy_types:
     assert.fail("Invalid Strategy type  \""+strategytype+"\", it must be OnDelete or RollingUpdate")
   end
   return strategy_type
end
```

and remove from all overlays

File: pkg/v1/providers/infrastructure-aws/v1.2.0/ytt/overlay.yaml [hunk 3]
```diff
@@ -9,6 +9,16 @@
...
+#@ strateg_type = ["OnDelete", "RollingUpdate"]
+
+
+#@ def verify_and_configure_strategy(strategytype):
+#@   if strategytype not in strateg_type:
+#@     assert.fail("Invalid Strategy type  \""+strategytype+"\", it must be OnDelete or RollingUpdate")
+#@   end
+#@   return strategytype
+#@ end
+
```
Comment (randomvariable) [inline pkg/v1/providers/infrastructure-aws/v1.2.0/ytt/overlay.yaml:21]: Remove and add to helpers.star

File: pkg/v1/providers/infrastructure-aws/v1.2.0/ytt/overlay.yaml [hunk 4]
```diff
@@ -188,6 +198,8 @@ spec:
   replicas: #@ data.values.WORKER_MACHINE_COUNT_0
   selector:
     matchLabels: null
+  strategy:
+    type: #@ verify_and_configure_strategy(data.values.WORKER_ROLLOUT_STRATEGY)
```
Comment (randomvariable) [inline pkg/v1/providers/infrastructure-aws/v1.2.0/ytt/overlay.yaml:202]:
```suggestion
    type: #@ verify_and_configure_machine_deployment_rollout_strategy(data.values.WORKER_ROLLOUT_STRATEGY)
```

`verify_and_configure_strategy` doesn't have the context of where it's happening.

File: pkg/v1/providers/infrastructure-azure/v1.2.1/ytt/overlay.yaml [hunk 5]
```diff
@@ -249,6 +259,8 @@ metadata:
 spec:
   clusterName: #@ data.values.CLUSTER_NAME
   replicas: #@ data.values.WORKER_MACHINE_COUNT_0
+  strategy:
+    type: #@ verify_and_configure_strategy(data.values.WORKER_ROLLOUT_STRATEGY)
```
Comment (randomvariable) [inline pkg/v1/providers/infrastructure-azure/v1.2.1/ytt/overlay.yaml:263]:
```suggestion
    type: #@ verify_and_configure_machine_deployment_rollout_strategy(data.values.WORKER_ROLLOUT_STRATEGY)
```

See previous

File: pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/base-template.yaml [hunk 6]
```diff
@@ -61,6 +61,8 @@ spec:
           networkName: '${ VSPHERE_NETWORK }'
       numCPUs: ${ VSPHERE_CONTROL_PLANE_NUM_CPUS }
       resourcePool: '${ VSPHERE_RESOURCE_POOL }'
+      pciDevices: ""
+      customVMXKeys: ""
```
Comment (randomvariable) [inline pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/base-template.yaml:65]: This is invalid for the CRD

File: pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/base-template.yaml [hunk 7]
```diff
@@ -87,6 +89,8 @@ spec:
           networkName: '${ VSPHERE_NETWORK }'
       numCPUs: ${ VSPHERE_WORKER_NUM_CPUS }
       resourcePool: '${ VSPHERE_RESOURCE_POOL }'
+      pciDevices: ""
+      customVMXKeys: ""
```
Comment (randomvariable) [inline pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/base-template.yaml:93]: Invalid types

File: pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml [hunk 8]
```diff
@@ -1,14 +1,60 @@
...
 #@ load("@ytt:data", "data")
 #@ load("@ytt:regexp", "regexp")
+#@ load("@ytt:assert", "assert")
 
 
+#@ valid_pci_deviceid = ["1EB8"]
+
+#@ valid_pci_vendorid = ["10DE"]
+
+#@ strateg_type = ["OnDelete", "RollingUpdate"]
```
Comment (randomvariable) [inline pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml:11]: Dedupe and remove

File: pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml [hunk 9]
```diff
@@ -1,14 +1,60 @@
 #@ load("@ytt:overlay", "overlay")
 #@ load("@ytt:data", "data")
 #@ load("@ytt:regexp", "regexp")
+#@ load("@ytt:assert", "assert")
 
 
+#@ valid_pci_deviceid = ["1EB8"]
```
Comment (randomvariable) [inline pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml:7]:
```suggestion
#@ valid_pci_deviceids = ["1EB8"]
```

File: pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml [hunk 10]
```diff
@@ -1,14 +1,60 @@
...
+
+#@ valid_pci_vendorid = ["10DE"]
```
Comment (randomvariable) [inline pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml:9]:
```suggestion
#@ valid_pci_vendorids = ["10DE"]
```

File: pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml [hunk 11]
```diff
@@ -1,14 +1,60 @@
...
 #@ load("lib/validate.star", "validate_configuration")
 #@ load("@ytt:yaml", "yaml")
 #@ validate_configuration("vsphere")
 
+#@ def verify_and_configure_strategy(strategytype):
+#@   if strategytype not in strateg_type:
+#@     assert.fail("Invalid Strategy type  \""+strategytype+"\", it must be OnDelete or RollingUpdate")
+#@   end
+#@   return strategytype
+#@ end
```
Comment (randomvariable) [inline pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml:24]: Dedupe and put into helpers.star

File: pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml [hunk 12]
```diff
@@ -226,6 +280,14 @@ spec:
           dhcp4: true
           #@ end
       numCPUs: #@ data.values.VSPHERE_WORKER_NUM_CPUS
+#@ if data.values.VSPHERE_WORKER_PCI_DEVICES != "" and data.values.VSPHERE_WORKER_PCI_DEVICES != None:
+      #@overlay/replace
```
Comment (randomvariable) [inline pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml:284]:
```suggestion
      #@overlay/match missing_ok=True
```

File: pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml [hunk 13]
```diff
@@ -226,6 +280,14 @@ spec:
...
+      pciDevices: #@ get_pci_devices(data.values.VSPHERE_WORKER_PCI_DEVICES, data.values.VSPHERE_IGNORE_PCI_DEVICES_ALLOW_LIST)
+#@ end
+#@ if data.values.VSPHERE_WORKER_CUSTOM_VMX_KEYS != "" and data.values.VSPHERE_WORKER_CUSTOM_VMX_KEYS != None:
+      #@overlay/replace
```
Comment (randomvariable) [inline pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml:288]:
```suggestion
      #@overlay/match missing_ok=True
```

**Review by randomvariable — CHANGES_REQUESTED** [review]
Thanks for this. Looking good so far!

There's some changes to be made to make the code more explicit, dedupe and fix some typos. 

Almost there.

File: pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml (hunk 13 above)
Comment (NEWBEE1) [reply pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml:288]: Hi @randomvariable  , We need to use "replace" for get_pci_devices() functionality to work . 
what is the need for using match ? Please can you explain this suggestion ?

File: pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml [hunk 14]
```diff
@@ -1,14 +1,60 @@
...
+
+
+#@ def get_pci_devices(pciDevicesString, pciIgnorepciDeviceValidation):
+#@   for val in pciDevicesString.split(','):
+#@     if len(val) <= 0:
+#@        assert.fail("given labels string \""+pciDevicesString+"\" must be in the  \"vendorID:deviceID,vendorID:deviceID,..\" format ")
+#@      end
+#@     kv = val.split(':')
+#@     if len(kv) != 2:
+#@        assert.fail("given labels string \""+pciDevicesString+"\" must be in the  \"vendorID:deviceID,vendorID:deviceID,..\" format ")
```
Comment (sathyanarays) [inline pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml:35]: Can this be merged into a single `if` block?

File: pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml [hunk 15]
```diff
@@ -1,14 +1,60 @@
...
+#@      end
+#@     if not pciIgnorepciDeviceValidation and (kv[0] not in valid_pci_vendorid or kv[1] not in valid_pci_deviceid):
```
Comment (sathyanarays) [inline pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml:37]: This looks good only when a device Id is universally unique. If they are not unique, this validation may go wrong. For example, let's say we have vendors V1 & V2; V1 has device with Ids D1, D2 & V2 has device with Ids D2 & D3. If we are going to support only V1's D2 & V2's D3, this logic will allow V2's D2 as well.

File: pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml (hunk 14 above)
Comment (NEWBEE1) [reply pkg/v1/providers/infrastructure-vsphere/v1.1.0/ytt/overlay.yaml:35]: It will not work

**Review by randomvariable — APPROVED** [review]
/lgtm

**Pull request closed without merging**